### PR TITLE
Support for multiple backends

### PR DESCRIPTION
**Pull request by fushinari**
Just starting this here so that I remind myself to get back around to it when I can. I think the plan should be to start with #112 (using the [LibreTranslate.com](https://libretranslate.com) instance initially) and then decide where we go from there.

Some decisions need to be made.
- [x] ~Do we simply clear history each time backend is switched or do we maintain separate histories for each backend?~ We clear.
- [x] ~On a similar note, how do we handle saving language preferences? Should we just use some defaults that are available in all services and then reset every time backend is switched or should we maintain separate defaults and separate saved languages?~ Saved per-backend.
- [x] ~What do we do for the UI for switching? Just an HdyComboRow? How badly or how well does it handle long lists? (if we end up implementing too many backends)~ Needs more investigation but let's stick with HdyComboRow for now.
- [x] ~How many backends should be available? How many is too many?~ Again, just start with what we can. So, probably just #112 for now.

To-do:
- [x] Initial implementation
- [x] Add LibreTranslate support
- [x] Per-backend settings and language lists
- [x] Dependency updates
- [x] Show a loading view on translator init
- [x] Support for switching LibreTranslate instances
- [x] Error handling for Translator initialization failure and support for retrying initialization
- [x] Basic instance URL validation support
- [x] Decide on a design for when URL validation fails
- [x] Some cleanup
- [x] Update Search Provider

### CODE REVIEW

**Comment by rafaelmardojai** [comment]
I think we need a `Translator` class witch the UI talks to, and this class manages the history and calls the selected provider.

>  Do we simply clear history each time backend is switched or do we maintain separate histories for each backend?

For now we can clear it each time backend is switched.

> On a similar note, how do we handle saving language preferences? Should we just use some defaults that are available in all services and then reset every time backend is switched or should we maintain separate defaults and separate saved languages?

Defaults and recent languages should be saved per backend I think.

> What do we do for the UI for switching? Just an HdyComboRow? How badly or how well does it handle long lists? (if we end up implementing too many backends)

I don't think we are going to have many backends for now. I don't know about the performance of `HdyComboRow`, but soon we will port to GTK4/libadwaita and I hope things will be better.

>  How many backends should be available? How many is too many?

For now, looks like `libretranslate.com / Argos Translate` is the only backend we can implement. We currently don't know if #47 or #114 are viable.

**Comment by fushinari** [comment]
> I think we need a Translator class witch the UI talks to, and this class manages the history and calls the selected provider.

My current planned approach is that every specific translation service's class will be inherited from a `BaseTranslator` class and they will be called based on a gsettings value at initialization. The results of `translate()` should either be an instance of the `Translation` class or something that inherits from it.

The history will be stored in the class instance that's initiated, which will mean that the history clearing takes place automatically.

The basic idea is to make everything return data similar to the py-googletrans API. For now, we're importing `GTranslator` directly but my plan is to have everything be automatically listed, called, etc. Kinda like the ["servers" implementation in Komikku](https://gitlab.com/valos/Komikku/-/tree/master/komikku/servers).

Is the plan fine as is or should we make changes?

**Comment by rafaelmardojai** [comment]
> Is the plan fine as is or should we make changes?

Looks good to me! Go ahead.

**Comment by fushinari** [comment]
There's a lot more work to do. Adding and updating dependencies (make httpx explicit, add langdetect), making the translators switchable (right now, I'm hardcoding for testing), fixing the visibility of the voice button when switching translators,  etc. But I'm taking a break from this for maybe two days. @rafaelmardojai if you want to commit in the meantime, go ahead. If not, no worries, I'll be back soon anyway. :)

**Comment by fushinari** [comment]
I'm too tired to update flatpak dependencies now, but it's finally in an actually working state. Just barely but it's working. It's now possible to switch between LibreTranslate (the astian.org instance) and Google Translate.

**Comment by fushinari** [comment]
Flatpak dependencies have been updated, so anyone who wants to check this out can try the latest CI build and it'll hopefully work well.

@rafaelmardojai could you check for any glaring bugs?

I'll be taking a long break now (about a month) due to exams.

**Comment by fushinari** [comment]
To-do:
- [ ] Per-backend Settings
- [ ] Some cleanup

**Comment by rafaelmardojai** [comment]
I'm going to test and try to finish the remaining tasks if I find the time.

**Comment by rafaelmardojai** [comment]
Tested, and To-do list updated .

Now we have per-backend settings and a loading view for backends init. 

Please test when you have time.

**Comment by fushinari** [comment]
I tested and it works really well! So, I've added in support for changing the LibreTranslate instance (which is a must-have according to the author of LibreTranslate) but I introduced a bug along the way where an invalid address will get us stuck on the loading page because connection fails. So, there are a few ways to fix this:

- Check validity of instance address before saving using a test button that will check, handle errors and fall back to defaults or save the new address depending on if it's valid.
- Just do error handling and send notifications to the user saying it's misconfigured

I think the first method might be a cleaner one just because it lets the user know beforehand if something is wrong. Gonna sleep for now and I really need to get started with studies :sweat_smile: (I've kinda just been slacking away still)

**Comment by rafaelmardojai** [comment]
I have made some progress. Now we have an initial "backend loading error view", so we can handle any error. And also added a new work flow to edit the backend instance that will allows us to validate the URI before saving, also it will reload the backend every time you change the instance because it could have different languages.

Please test, as a break of the hard study 😅.

**Comment by rafaelmardojai** [comment]
How do you think we should validate the address? Maybe the backend should have a `validate_instance_url` function?

**Comment by fushinari** [comment]
> Please test, as a break of the hard study :sweat_smile:.

Thank you! It's been hard... I tested it and it works really well from what I see!

> How do you think we should validate the address? Maybe the backend should have a `validate_instance_url` function?

That is what I had in mind too, but then again, how would we check if an instance is valid without calling call the functions? In the case of LibreTranslate, maybe we can just check the [frontend settings](https://libretranslate.com/docs/#/frontend/get_frontend_settings).

There's one more issue with supporting LibreTranslate instances. Character limits can be different between instances. The LibreTranslate.com instance limits it to 5000 characters but the Astian instance allows unlimited characters. I was thinking of allowing character limits to be per-backend, but then again, maybe it will lead to abuse and a hard in-app limit is good to have?

**Comment by rafaelmardojai** [comment]
I didn't see your commits and I had some trouble 😂. I have added an initial verification code.

It's 2AM here, so probably I don't know what I'm doing. I'm going to sleep now.

File: data/resources/preferences.ui [hunk 1]
```diff
@@ -56,6 +56,108 @@
...
+                          </object>
+                        </child>
+                        <child>
+                          <object class="GtkButton" id="backend_instance_edit">
+                            <property name="visible">True</property>
+                            <property name="can-focus">True</property>
+                            <property name="receives-default">True</property>
+                            <property name="tooltip-text" translatable="yes">Edit</property>
+                            <style>
+                              <class name="image-button"/>
```
Comment (fushinari) [inline data/resources/preferences.ui:96]: Do we want `image-button`?

With `image-button`:
image

Without it:
image

Also, using `set_image` apparently also sets `image-button`. And using `image-button` with `suggested-action` makes the icon invisible. At first, I just thought "GTK Bug :/" and worked around it, but it's apparently intended behavior. At least one part of it is.

Comment (fushinari) [reply data/resources/preferences.ui:96]: This is the bug I'm talking about:
image

**Comment by fushinari** [comment]
> I didn't see your commits and I had some trouble :joy:. 

Ah, sorry for that, just do `git pull --rebase` each time before pushing and we'll be fine I think. :sweat_smile: 

I implemented a sort of error popup. But design needs some fixing. Actually, I'm not even sure that using a popover is good design. I'm quite bad at design in general, so I'm going to leave that up to you. :sweat_smile: 

Current form:


Possible alternatives:

**Comment by fushinari** [comment]
I've also moved the to-do list to the original post and added all the stuff we've already completed just to keep track of all of it. :)

**Comment by rafaelmardojai** [comment]
> I implemented a sort of error popup. But design needs some fixing. Actually, I'm not even sure that using a popover is good design. I'm quite bad at design in general, so I'm going to leave that up to you. 

I prefer the error borders only in the entry.

**Comment by fushinari** [comment]
> I prefer the error borders only in the entry.

Yep, done!

**Comment by rafaelmardojai** [comment]
Nice. 

Now comes the fun part of updating the search provider... 😂

**Comment by fushinari** [comment]
Oh yeah... Before that, we'd also have to disable the backend switcher right?

**Comment by rafaelmardojai** [comment]
> Before that, we'd also have to disable the backend switcher right?

What do you mean?

**Comment by fushinari** [comment]
During validation, don't we want the backend switcher to also be insensitive?

**Comment by fushinari** [comment]
2928a51 is what I meant

**Comment by fushinari** [comment]
Wow... what do we do for the search provider?

File: data/resources/preferences.ui (hunk 1 above)
Comment (fushinari) [reply data/resources/preferences.ui:96]: @rafaelmardojai what do you think about this?

**Comment by rafaelmardojai** [comment]
> Wow... what do we do for the search provider?

Mmm, probably the same as now but checking the backend setting and importing the dialect translators module.

**Comment by fushinari** [comment]
Ah, for some reason, I assumed that the dialect package would be inaccessible from the search provider.

File: data/resources/preferences.ui (hunk 1 above)
Comment (fushinari) [reply data/resources/preferences.ui]: Fixed in e508b90

**Comment by fushinari** [comment]
I think it's finally ready. Bit more testing and I think we can merge.

**Comment by fushinari** [comment]
@rafaelmardojai do you have any pending changes? I want to rebase if possible

**Comment by rafaelmardojai** [comment]
No sorry, feel free to rebase. 

We also need to add `preferences.py` to `POTFILES`

**Comment by fushinari** [comment]
Oof... that was not the result I intended...

**Comment by fushinari** [comment]
@rafaelmardojai If I rebase, all your commits are going to show up as "Authored by you and committed by me". Should I just merge instead?

**Comment by rafaelmardojai** [comment]
Yeah no problem

**Comment by fushinari** [comment]
@rafaelmardojai I'm having a very stable experience with this so far.

What do you think we should do next? Call on translators, update and release? Or should we wait, test a bit more, and then release? Also, did you get to test the search provider? Oh, also, do you think the preferences' descriptions right now are okay?

So many questions :laughing: 

On another note, I'm really glad we've finally wrapped up work on the longest standing request so far! LibreTranslate support is complete even with support for using any instance and the framework is there for more translation services! I'm really happy with this!

**Comment by rafaelmardojai** [comment]
I want to put a label in the loading view, then we can call translators for `1.2.0`?

Preferences' descriptions  looks good to me.

**Comment by fushinari** [comment]
Yep, sounds good, go on. :)

**Comment by TheOPtimal** [comment]
I wanted to mention Argos Translate, which LibreTranslate actually uses on the backend. It runs completely offline, and is available as a [python library.](https://github.com/argosopentech/argos-translate)

**Comment by fushinari** [comment]
@TheOPtimal I did consider that but then, I saw that it requires trained models to be manually downloaded by the user and other manual intervention. I also realized that a user of Argos Translate could easily setup a local instance of LibreTranslate if he wishes to. So, to make it as much of an out-of-the-box experience as possible while still giving the user enough power, I ended up deciding on this compromise.

**Comment by fossdd** [comment]
Awesome change!

**Pull request merged**